### PR TITLE
HAL: use hal.serial(n) to access serial ports

### PR DESCRIPTION
**Pull request by tridge**
This removes the use of hal.uartX for uart access in all libraries and vehicle code. It is replaced with hal.serial(n), which matches the SERIALn_ numbering and also matches the SERIAL_ORDER in hwdef.dat

The hal.uartX variables are still inside the hal as private members of the hal, and a future PR will eliminate those, but this removes the exposure of developers to the very odd hal.uartX ordering immediately, which has caused considerable confusion
The flash cost for omnibusf4pro plane of this change is 108 bytes, and saves 4 bytes of ram

### CODE REVIEW

**Review by davidbuzz — APPROVED** [review]
reviewed, only saw that one typo in the comment.

File: libraries/AP_HAL/examples/UART_chargen/UART_chargen.cpp [hunk 1]
```diff
@@ -41,14 +41,14 @@ void setup(void)
...
-    // uart = hal.uartB;   // 1st GPS
+    // uart = hal.serial(3);   // 1st GPS
 #if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
-    uart = hal.uartC;  // telemetry 1
+    uart = hal.serial(1);  // telemetry 1
 #endif
-    // uart = hal.uartD;  // telemetry 2
-    // uart = hal.uartE;  // 2nd GPS
+    // uart = hal.serial(2);  // telemetry 2
+    // uart = hal.serial(3);  // 2nd GPS
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL/examples/UART_chargen/UART_chargen.cpp]: fixed

**Review by magicrub — APPROVED** [review]
I tested this on another branch that I'll be PRing right after this goes in. This simplified it a little bit. I look forward to this going in!

**Comment by magicrub** [comment]
I tested PR https://github.com/ArduPilot/ardupilot/pull/16041 on this branch.

**Comment by magicrub** [comment]
@Hwurzburg this will need some wiki changes.

**Comment by magicrub** [comment]
I just retested this on https://github.com/ArduPilot/ardupilot/pull/16041 and decided to just change that PR to be on top of this one. That PR is just a single commit.

**Comment by Hwurzburg** [comment]
> @Hwurzburg this will need some wiki changes.

@magicrub Tom, what changes and where?

**Comment by magicrub** [comment]
Anything that says uartA, B,C,D,E...

**Comment by Hwurzburg** [comment]
> Anything that says uartA, B,C,D,E...
only mentions in wiki are a few tables and using SITL:
did sim_vehicle.py change its arguments from like _sim_vehicle.py -A "--uartC=uart:/dev/ttyS15"_? if so, what would be the proper syntax now?

**Comment by magicrub** [comment]
Good question. I guess we'll have to test

**Comment by Hwurzburg** [comment]
@magicrub still awaiting test results...

**Comment by magicrub** [comment]
Thx. I haven't specifically tested this but if it wasn't working correctly we would have known by now by other people complaining.

**Pull request merged**